datafusion/functions/src/string/starts_with.rs

```diff
@@ -53,16 +76,18 @@ impl Default for StartsWithFunc {
 impl StartsWithFunc {
     pub fn new() -> Self {
         use DataType::*;
+
+        let string_types = vec![Utf8, LargeUtf8, Utf8View];
+        let mut type_signatures = vec![];
+
+        for left in &string_types {
```

Comment:
Not sure if there's a better way to do this? Maybe I should just list them out?

Reply:
This is a interesting idea, i liked it compared to earlier implementation.

Reply:
I don't think the underlying kernel supports different string types (you can't actually call it with (`Utf8`, `LargeUtf8`) so the existing function actually has a bug

In fact here is reproducer showing the probelm on main
```sql
DataFusion CLI v40.0.0
> create table foo as values (arrow_cast('foo', 'Utf8'), arrow_cast('bar', 'LargeUtf8'));
0 row(s) fetched.
Elapsed 0.046 seconds.

> select * from foo;
+---------+---------+
| column1 | column2 |
+---------+---------+
| foo     | bar     |
+---------+---------+
1 row(s) fetched.
Elapsed 0.010 seconds.

> select starts_with(column1, column2) from foo;
Internal error: could not cast value to arrow_array::array::byte_array::GenericByteArray<arrow_array::types::GenericStringType<ii32>>.
This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

So in other words, I think the correct signature is

```rust
            signature: Signature::one_of(
                vec![
                    Exact(vec![Utf8, Utf8]),
                    Exact(vec![LargeUtf8, LargeUtf8]),
                    Exact(vec![Utf8View, Utf8View]),
                ],
                Volatility::Immutable,
            ),
        }
```

Reply:
Thanks for this and link to the tests -- very helpful!

I made this update along with simplifying the arrow starts_with call (https://github.com/apache/datafusion/pull/11787#discussion_r1702177110), and I'm noticing that `STARTS_WITH(column1_utf8view, 'foo')` results in a cast of the first column to utf8 while keeping the second one utf8, but I think we'd want to keep the first argument as a utf8view while coercing the second argument to utf8view? It feels like calling `starts_with` with a second argument as a scalar is common.

For example, with the changes, this test passes:

```
query TT
EXPLAIN SELECT
  STARTS_WITH(column1_utf8view, 'foo') as c,
  STARTS_WITH(column1_utf8view, column2_utf8view) as c2
FROM test;
----
logical_plan
01)Projection: starts_with(CAST(test.column1_utf8view AS Utf8), Utf8("foo")) AS c, starts_with(test.column1_utf8view, test.column2_utf8view) AS c2
02)--TableScan: test projection=[column1_utf8view, column2_utf8view]
```

Reply:
Yes I agree figuring out how to coerce the constant to Utf8view is important

I think the coercion logic just goes linearly through the signature and stops at the first type that works.

Thus I wonder if you can make `Utf8View` preferred by listing the types in a different order and put the `Utf8View` signature at the first option?  Like

```rust
            signature: Signature::one_of(
                vec![
                    // make `Utf8View` first
                    Exact(vec![Utf8View, Utf8View]),
                    Exact(vec![Utf8, Utf8]),
                    Exact(vec![LargeUtf8, LargeUtf8]),
                ],
                Volatility::Immutable,
            ),
        }
```

Reply:
Still seems to be happening with this latest push 🤔 ([222263c](https://github.com/apache/datafusion/pull/11787/commits/222263c152c002243e4bed18147758ce0365bb12)).

I'll try to poke around and see if I can understand what's happening.

Reply:
I think it needed this update in the argument coercion logic... before this even with Utf8View being preferred in the signature it wasn't recognized in the type to/from statement: https://github.com/apache/datafusion/pull/11787/files#diff-6e1fb265597317a8256c60670ff3ea7be6896b2df1199a40ca79419ce29b4ce3